### PR TITLE
readme and log example

### PR DESCRIPTION
**Pull request by mcuadros**
https://gist.github.com/mcuadros/21c86bc1c32952e3ae65ac0d813e9c37

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,28 +1,19 @@
...
+A low level and highly extensible git implementation in **pure Go**. 
 
-We have been following the open/close principle in its design to facilitate extensions.
+*go-git* aims to reach the completeness of [libgit2](https://libgit2.github.com/) or [jgit](http://www.eclipse.org/jgit/), nowadays covers the **majority** of the plumbing **read operations** and **some** of the main **write operations**, but lacks of the main porcelain operations such as merges. 
 
-*go-git* does not claim to be a replacement of [git2go](https://github.com/libgit2/git2go) as its approach and functionality is quite different.
+**highly extensible**, we have been following the open/close principle in its design to facilitate extensions, mainly focusing the efforts on the persistence of the objects.
 
-### ok, but why? ...
+### ... but this is production ready?
```
Comment (vmarkovtsev) [inline README.md]: but is this production ready?

File: README.md [hunk 2]
```diff
@@ -1,28 +1,19 @@
...
+# go-git [![GoDoc](https://godoc.org/gopkg.in/src-d/go-git.v4?status.svg)](https://godoc.org/gopkg.in/src-d/go-git.v4) [![Build Status](https://travis-ci.org/src-d/go-git.svg)](https://travis-ci.org/src-d/go-git) [![codecov.io](https://codecov.io/github/src-d/go-git/coverage.svg)](https://codecov.io/github/src-d/go-git) [![codebeat badge](https://codebeat.co/badges/b6cb2f73-9e54-483d-89f9-4b95a911f40c)](https://codebeat.co/projects/github-com-src-d-go-git)
 
-A low level and highly extensible git client library for **reading** repositories from git servers.  It is written in Go from scratch, without any C dependencies.
+A low level and highly extensible git implementation in **pure Go**. 
 
-We have been following the open/close principle in its design to facilitate extensions.
+*go-git* aims to reach the completeness of [libgit2](https://libgit2.github.com/) or [jgit](http://www.eclipse.org/jgit/), nowadays covers the **majority** of the plumbing **read operations** and **some** of the main **write operations**, but lacks of the main porcelain operations such as merges.
```
Comment (Serabe) [inline README.md]: I think _lack_ only has _of_ after when used as a noun.

File: README.md [hunk 3]
```diff
@@ -1,28 +1,19 @@
...
+*go-git* aims to reach the completeness of [libgit2](https://libgit2.github.com/) or [jgit](http://www.eclipse.org/jgit/), nowadays covers the **majority** of the plumbing **read operations** and **some** of the main **write operations**, but lacks of the main porcelain operations such as merges. 
 
-*go-git* does not claim to be a replacement of [git2go](https://github.com/libgit2/git2go) as its approach and functionality is quite different.
+**highly extensible**, we have been following the open/close principle in its design to facilitate extensions, mainly focusing the efforts on the persistence of the objects.
 
-### ok, but why? ...
+### ... but this is production ready?
 
-At [source{d}](http://sourced.tech) we analyze almost **all** the public open source contributions made to git repositories in the world.
+The master branch, represents the `v4` of the library, currently actively development, which is planned to be released in early 2017.
```
Comment (vmarkovtsev) [inline README.md]: The master branch represents the `v4` of the library, it is currently under active development and is planned to be released in early 2017.

File: README.md [hunk 4]
```diff
@@ -1,28 +1,19 @@
...
 
-A low level and highly extensible git client library for **reading** repositories from git servers.  It is written in Go from scratch, without any C dependencies.
+A low level and highly extensible git implementation in **pure Go**. 
 
-We have been following the open/close principle in its design to facilitate extensions.
+*go-git* aims to reach the completeness of [libgit2](https://libgit2.github.com/) or [jgit](http://www.eclipse.org/jgit/), nowadays covers the **majority** of the plumbing **read operations** and **some** of the main **write operations**, but lacks of the main porcelain operations such as merges. 
 
-*go-git* does not claim to be a replacement of [git2go](https://github.com/libgit2/git2go) as its approach and functionality is quite different.
+**highly extensible**, we have been following the open/close principle in its design to facilitate extensions, mainly focusing the efforts on the persistence of the objects.
```
Comment (Serabe) [inline README.md]: Lack of verb. _It is highly extensible_.

File: README.md (hunk 1 above)
Comment (Serabe) [inline README.md]: _is this production ready?_

File: README.md [hunk 5]
```diff
@@ -1,28 +1,19 @@
...
 
-We want to extract detailed information from each GitHub repository, which requires downloading repository packfiles and analyzing them: extracting their code, authors, dates and the languages and ecosystems they use.  We are also interested in knowing who contributes to what, so we can tell top contributors from the more casual ones.
+If you are looking for a production ready version, please take a look to the [`v3`](https://github.com/src-d/go-git/tree/v3) which is being use in production at [source{d}](http://sourced.tech) since August 2015 to analyze all GitHub public repositories (i.e. 16M of repositories).
 
-You can obtain all this information using the standard `git` command running over a local clone of a repository, but this simple solution does not scale well over millions of repositories: we want to avoid having local copies of the unpacked repositories in a regular file system; *go-git* allows us to work with an in-memory representation of repositories instead.
+ We recommend use the `v4` to develop new projects, since it include a lot of new functionality and with a more git idiomatic API.
```
Comment (vmarkovtsev) [inline README.md]: We recommend to use the `v4` to develop new projects since it includes much new functionality and has more idiomatic git API.

File: README.md [hunk 6]
```diff
@@ -90,100 +84,13 @@ Date:   2015-12-11 17:57:10 +0100 +0100
...
-directory for a simple filesystem implementation and usage.
-
-Wrapping
---------
-
-go-git can be wrapped into any language which supports shared library interop.
-[Python wrapper](https://github.com/src-d/gypogit) already exists.
-This is provided by "cshared" [cgo](https://golang.org/cmd/cgo/) files which can be built
-with `go build -o libgogit.so -buildmode=c-shared github.com/src-d/go-git/cshared`.
+Take note that the functions `CheckIfError` and `Info` are from the [examples package](https://github.com/src-d/go-git/blob/master/examples/common.go#L17) just to be used in the examples.
```
Comment (vmarkovtsev) [inline README.md]: Please note  that the functions `CheckIfError` and `Info` used in the examples were taken from the ...

File: README.md [hunk 7]
```diff
@@ -90,100 +84,13 @@ Date:   2015-12-11 17:57:10 +0100 +0100
...
 
-Acknowledgements
-----------------
 
-The earlier versions of the [packfile reader](https://godoc.org/gopkg.in/src-d/go-git.v4/formats/packfile) are based on [git-chain](https://github.com/gitchain/gitchain/blob/master/git/pack.go), project done by [@yrashk](https://github.com/yrashk)
+Contribute
+----------
 
+If you're interested in being a contributor and want to get involved in developing go-git, the easiest way is open a [issue](https://github.com/src-d/go-git/issues) telling us what missing functionality you want to work in, and we will guide you about how implement it.
```
Comment (vmarkovtsev) [inline README.md]: involved in go-git development

the easiest way is to open an issue ... which describes what missing functionality you want to work on and we will guide you how to implement it best.

File: README.md (hunk 3 above)
Comment (Serabe) [inline README.md]: * [ ] Remove comma after _master branch.
* [ ] _currently under active development_ or _currently actively developed_. Though I don't like two adverbs one after the other.

I would rephrase:

> The master branch represents the `v4` of the library planned to be released in early 2017.

The main idea is together, removes pronouns, and the library being actively developed is implied in the _release early 2017_.

File: README.md [hunk 8]
```diff
@@ -1,28 +1,19 @@
...
+**highly extensible**, we have been following the open/close principle in its design to facilitate extensions, mainly focusing the efforts on the persistence of the objects.
 
-### ok, but why? ...
+### ... but this is production ready?
 
-At [source{d}](http://sourced.tech) we analyze almost **all** the public open source contributions made to git repositories in the world.
+The master branch, represents the `v4` of the library, currently actively development, which is planned to be released in early 2017.
 
-We want to extract detailed information from each GitHub repository, which requires downloading repository packfiles and analyzing them: extracting their code, authors, dates and the languages and ecosystems they use.  We are also interested in knowing who contributes to what, so we can tell top contributors from the more casual ones.
+If you are looking for a production ready version, please take a look to the [`v3`](https://github.com/src-d/go-git/tree/v3) which is being use in production at [source{d}](http://sourced.tech) since August 2015 to analyze all GitHub public repositories (i.e. 16M of repositories).
```
Comment (Serabe) [inline README.md]: * [ ] _being used_.
* [ ] _16M repositories_.

File: README.md (hunk 5 above)
Comment (Serabe) [inline README.md]: * [ ] _recommend using_ or _recommend the use_.
* [ ] _it includes_
* [ ] _and provides a more idiomatic git API_ or _and provides a more git-like API_.

File: README.md [hunk 9]
```diff
@@ -37,36 +28,39 @@ go get -u gopkg.in/src-d/go-git.v4/...
...
-if err != nil {
-	panic(err)
-}
+// We instance an in-memory git repository
+r := git.NewMemoryRepository()
 
-if err := r.PullDefault(); err != nil {
-	panic(err)
-}
+// Clone the given repository, creating the remote, the local branches
```
Comment (Serabe) [inline README.md]: It is a bit strange to comment a line of code by commenting the line above it and using part of it as part of the comment.

File: README.md [hunk 10]
```diff
@@ -37,36 +28,39 @@ go get -u gopkg.in/src-d/go-git.v4/...
...
 
-iter, err := r.Commits()
-if err != nil {
-	panic(err)
-}
-defer iter.Close()
+err := r.Clone(&git.CloneOptions{URL: "https://github.com/src-d/go-siva"})
+CheckIfError(err)
+
+// Getting the HEAD history from HEAD, just like does:
```
Comment (Serabe) [inline README.md]: In the previous comment you used the imperative voice to define the action the code is doing.

File: README.md [hunk 11]
```diff
@@ -37,36 +28,39 @@ go get -u gopkg.in/src-d/go-git.v4/...
...
-		if err == io.EOF {
-			break
-		}
+// ... retrieving the branch being pointed by HEAD
+ref, err := r.Head()
+CheckIfError(err)
 
-		panic(err)
-	}
+// ... retrieving the commit object
```
Comment (Serabe) [inline README.md]: Same as above.

File: README.md [hunk 12]
```diff
@@ -37,36 +28,39 @@ go get -u gopkg.in/src-d/go-git.v4/...
...
+commit, err := r.Commit(ref.Hash())
+CheckIfError(err)
 
-	fmt.Println(commit)
+// ... we retrieve the commit history
```
Comment (Serabe) [inline README.md]: Same as above, now using present.

File: README.md [hunk 13]
```diff
@@ -37,36 +28,39 @@ go get -u gopkg.in/src-d/go-git.v4/...
...
+history, err := commit.History()
+CheckIfError(err)
+
+// ... now just iterate over the commits, printing it
```
Comment (Serabe) [inline README.md]: Please, choose one time and use it in all the comments.

File: README.md (hunk 6 above)
Comment (Serabe) [inline README.md]: Move this notice before the code so people know it before reaching the code.

File: README.md (hunk 7 above)
Comment (Serabe) [inline README.md]: > If you are interested on contributing to go-git, open an [issue](https://github.com/src-d/go-git/issues) explaining which missing functionality you want to work in, and we will guide you through the implementation.
